### PR TITLE
Add promoted card to in-App marketplace page

### PR DESCRIPTION
**Pull request by corsonr**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

The search on `WooCommerce > Marketplace` page that shows product from the WooCommerce.com marketplace as cards. During previous re-design phase we added updated product cards in #30410.

This PR adds a new style of "promoted" product cards by:

- adding a new style card
- adding a top border to the card (the color is defined in the JSON feed)
- adding a colored button (the color is defined in the JSON feed)
- adding an automated logic that changes the button text color base on the button background-color

Closes Automattic/woocommerce.com#10885

### How to test the changes in this Pull Request:

- Check out this branch.
- Edit `includes/admin/class-wc-admin-addons.php`.
- On line 66, change `$featured = get_transient( 'wc_addons_featured' );` to bypass the transient check:

```
$featured = false; // TODO - REMOVE THIS
```

- On line 83, change `https://woocommerce.com/wp-json/wccom-extensions/2.0/featured' . $parameter_string` to point to our mock WCCOM featured API data: `https://gist.githubusercontent.com/corsonr/6ac9a36583dce3a2d2d9e97f4ddf7847/raw/60a38e52de705566bdb74b2059d6acc6a086379b/featured.json`.

```php
$raw_featured = wp_safe_remote_get(
	'https://gist.githubusercontent.com/andfinally/be6f6caae34f3f3c1c5de1df0ba30681/raw/featured.json',
	array(
		'headers'    => $headers,
		'user-agent' => 'WooCommerce Addons Page',
	)
);
```

the headers could have to be removed from the `wp_safe_remote_get` call.

- On line 142, change `'https://woocommerce.com/wp-json/wccom-extensions/1.0/search' . $parameters,` to point to another mock JSON endpoint: `https://gist.githubusercontent.com/andfinally/cdea0c85da10bc037dad6aaa8bf28d0e/raw/search.json`.

```php
$raw_extensions = wp_safe_remote_get(
	'https://gist.githubusercontent.com/andfinally/cdea0c85da10bc037dad6aaa8bf28d0e/raw/search.json',
	array( 'headers' => $headers )
);
```

*  View the Featured page of WooCommerce > Marketplace on your WooCommerce site `wp-admin/admin.php?page=wc-addons&section=_featured`.
* In the `Discover our favorites` group you should see the two new styles of card shown in this screenshot:

image

* View a category page like `wp-admin/admin.php?page=wc-addons&section=marketing-extensions` or search results page. You should see the two styles of card shown here at the top:

image

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Tweak - Add promoted cards styling to marketplace section. #30861

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by andfinally** [comment]
I'm currently having trouble getting the API request to work locally with my [Woo Core development environment site](https://github.com/Automattic/woocommerce-core-development-environment) – I'm getting `cURL error 7: Failed to connect to woocommerce.test port 443: Connection refused` despite having a filter to set the correct HTTP options. I'll try again later!

**Comment by corsonr** [comment]
@andfinally I'm testing from the test customer site if that helps 😉

File: includes/admin/class-wc-admin-addons.php [hunk 1]
```diff
@@ -33,7 +33,7 @@ public static function get_featured() {
 			}
 
 			$raw_featured = wp_safe_remote_get(
-				'https://woocommerce.com/wp-json/wccom-extensions/1.0/featured',
+				'https://woocommerce.test/wp-json/wccom-extensions/1.0/featured',
```
Comment (madeincosmos) [inline includes/admin/class-wc-admin-addons.php]: Let's leave this one out of production :)

File: assets/css/admin.scss [hunk 2]
```diff
@@ -726,13 +741,34 @@
...
+						justify-content: flex-end;
+						align-items: center;
+						padding: 3px 12px;
+						float: right;
+						height: 20px;
+						right: 24px;
+						top: 28px;
+						text-align: center;
+						background: #DCDCDE;
+						border-radius: 2px;
```
Comment (madeincosmos) [inline assets/css/admin.scss]: a minor thing, but it would be good to arrange these properties alphabetically like the linter suggests

**Review by madeincosmos — COMMENTED** [review]
Thank you for the PR @corsonr! Everything looks good to me except for the test site URL that accidentally made its way to the commit, we'll need to replace it with the original feed before merging. Other than that, the cards are looking great on all screen sizes 🌟

File: includes/admin/class-wc-admin-addons.php (hunk 1 above)
Comment (corsonr) [reply includes/admin/class-wc-admin-addons.php]: good catch!

File: includes/admin/class-wc-admin-addons.php [hunk 3]
```diff
@@ -305,6 +305,40 @@ public static function output_column_block( $block ) {
...
+	 * Handles the outputting of a promoted card.
+	 *
+	 * @param object $block Block data.
+	 */
+	public static function output_promoted_card( $block ) {
+		?>
+		<ul class="products">
+			<?php
+			foreach ( $block->items as $card ) {
+				if ( true === $card->full_width ) {
```
Comment (tgglv) [inline includes/admin/class-wc-admin-addons.php]: It seems that checking agains some kind of `isset` would be vital here. Currently with data from the PR's testing instructions I have to following notices:
image

Perhaps something like the following will fit us best (sorry, I'm syntactic sugar addict :yum:):
```php
$full_width = $card->full_width ?? false;
if ( true === $full_width ) {
    ...
```

File: includes/admin/class-wc-admin-addons.php (hunk 1 above)
Comment (tgglv) [reply includes/admin/class-wc-admin-addons.php]: It looks like `woocommerce.test` in the URL is still there. Please don't forget to use `woocommerce.com` before the final merge.

**Review by tgglv — COMMENTED** [review]
Thanks, @corsonr!

It works like a charm! 🍀 

I've made it as a comment rather than approval to ask you to apply Maria's and mine suggestions first.

**Review by madeincosmos — DISMISSED** [review]
Thank you for the fixes, everything looks good to me now on all screen sizes 🎉

**Comment by corsonr** [comment]
After merging @tgglv code I had to make a few tweaks to re-integrate some styling for the featured and promoted cards.

Now, in search results:

#### Featured Cards:


#### Promoted Cards


#### Regular Cards:

**Comment by andfinally** [comment]
For future reference, this is the structure of a "featured" and a "promoted" product that will be returned from the WCCOM API.

```json
{
    "title": "Stripe",
    "image": "https://woocommerce.com/wp-content/uploads/2020/06/Logo-Stripe-Dark-1.png",
    "excerpt": "Accept all major debit and credit cards as well as local payment methods with Stripe.",
    "link": "https://woocommerce.com/products/stripe/?utm_source=extensionsscreen&utm_medium=product&utm_campaign=wcaddons",
    "demo_url": "",
    "price": "&#36;0.00",
    "hash": "50bb7a985c691bb943a9da4d2c8b5efd",
    "slug": "woocommerce-gateway-stripe",
    "id": 18627,
    "rating": null,
    "reviews_count": null,
    "vendor_name": "WooCommerce",
    "vendor_url": "https://woocommerce.com/vendor/woocommerce/",
    "icon": "https://woocommerce.com/wp-content/uploads/woocommerce_uploads/2011/12/stripe-app-icon-7m1xi7.png",
    "label": "featured"
},
{
    "title": "WooCommerce Payments",
    "image": "https://woocommerce.com/wp-content/uploads/2020/06/Logo-Woo-Pay-Dark.png",
    "excerpt": "Securely accept payments, track cash flow, and manage recurring revenue from your dashboard \u2014 all without setup costs or monthly fees.",
    "link": "https://woocommerce.com/products/woocommerce-payments/?utm_source=extensionsscreen&utm_medium=product&utm_campaign=wcaddons",
    "demo_url": "",
    "price": "&#36;0.00",
    "hash": "8c6319ca-8f41-4e69-be63-6b15ee37773b",
    "slug": "woocommerce-payments",
    "id": 5278104,
    "rating": null,
    "reviews_count": null,
    "vendor_name": "WooCommerce",
    "vendor_url": "https://woocommerce.com/vendor/woocommerce/",
    "icon": "https://woocommerce.com/wp-content/uploads/woocommerce_uploads/2020/02/wcpay-icon-fv4tef.png",
    "primary_color": "#FFE320",
    "text_color": "#2c3338",
    "button": "View details",
    "label": "promoted"
},
```

**Review by andfinally — DISMISSED** [review]
Looking great, thank you! 👌

Apologies to other reviewers for the whitespace changes.

**Review by tgglv — DISMISSED** [review]
Thanks, @corsonr!

Changes look great and work like a charm 🍀 for the featured page, search results (for both: search by a term and category).

The only thing according to test instructions. I can't test with the GH gist (don't know why). Other than that I've used local WCCOM.

**Comment by andfinally** [comment]
Thanks Timur! In case needed, I've pasted the contents of the two gists into these a8c pastebins.

featured.json: 2a72a-pb/
search.json: 2a72b-pb/

**Review by tgglv — DISMISSED** [review]
Thanks, @corsonr!

The code looks great and works like a charm! 🍀 

Tested it once again after resolving conflicts with the `trunk`. Everything works as expected (for featured/promoted/regular product cards for both featured page and search).

**Comment by Konamiman** [comment]
I tested following the instructions and found that:

* `wp-admin/admin.php?page=wc-addons&section=_featured` shows one style of cards, not two:

image

* `wp-admin/admin.php?page=wc-addons&section=marketing-extensions` shows a blank page:

image

(that's still the case even if I select "All" in the dropdown)

**Comment by corsonr** [comment]
@Konamiman 

>wp-admin/admin.php?page=wc-addons&section=_featured shows one style of cards, not two

Instructions have been updated, `render_featured()` needs to be updated, since `get_featured()` is deprecated.

I'm getting correct results:

**Comment by andfinally** [comment]
Sorry Néstor, our test instructions referenced the wrong line numbers. The cards should now appear as expected.

**Comment by andfinally** [comment]
Hi @Konamiman, I've set the base to `release/5.9`, resolved the conflicts and tested again – this is good for review again, thank you!

**Review by Konamiman — APPROVED** [review]

**Comment by Konamiman** [comment]
Merged into `release/5.9` for inclusion in WooCommerce 5.9 RC. Needs to be cherry-picked into `trunk`.

**Comment by Konamiman** [comment]
Additional pull request to incorporate these changes in WooCommerce 6.0: https://github.com/woocommerce/woocommerce/pull/31090

**Pull request merged**